File: Semestrul 4/SDI/lab06/cs/Utils/DatabaseConfig.cs
using System;
using System.IO;
using System.Data.SQLite;
using System.Linq;
using System.Collections.Generic;

namespace cs.Utils;

public class DatabaseConfig
{
    private static DatabaseConfig instance;
    private static readonly object lockObject = new object();
    private static SQLiteConnection connection;

    private DatabaseConfig() { }

    public static DatabaseConfig GetInstance()
    {
        if (instance == null)
        {
            lock (lockObject)
            {
                if (instance == null)
                {
                    instance = new DatabaseConfig();
                }
            }
        }
        return instance;
    }

    public static SQLiteConnection GetConnection()
    {
        if (true)
        {
            try
            {
                string configPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "configs", "config.properties");

                if (!File.Exists(configPath))
                {
                    throw new IOException("Unable to find config.properties in the configs directory");
                }

                var properties = File.ReadAllLines(configPath)
                    .Where(line => !string.IsNullOrWhiteSpace(line) && !line.TrimStart().StartsWith("#"))
                    .Select(line => line.Split('=', 2))
                    .Where(parts => parts.Length == 2)
                    .ToDictionary(parts => parts[0].Trim(), parts => parts[1].Trim());

                if (!properties.TryGetValue("db.url", out string jdbcUrl) || string.IsNullOrEmpty(jdbcUrl))
                {
                    throw new IOException("db.url property not found or empty in config.properties");
                }

                string connectionString = $"Data Source={jdbcUrl};Version=3;";

                connection = new SQLiteConnection(connectionString);
                connection.Open();
            }
            catch (Exception ex)
            {
                throw new Exception("Failed to establish database connection", ex);
            }
        }

        return connection;
    }


    public void CloseConnection()
    {
        if (connection != null && connection.State == System.Data.ConnectionState.Open)
        {
            connection.Close();
        }
    }
}

File: Semestrul 4/SDI/lab11/java/TriathlonPersistence/src/main/java/triathlon/persistence/utils/DatabaseConfig.java
package triathlon.persistence.utils;

import java.io.IOException;
import java.io.InputStream;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;

public class DatabaseConfig {

    private static DatabaseConfig instance;

    private DatabaseConfig() {}

    public static DatabaseConfig getInstance() {
        if (instance == null) {
            synchronized (DatabaseConfig.class) {
                if (instance == null) {
                    instance = new DatabaseConfig();
                }
            }
        }
        return instance;
    }

    private static Connection connection;

    public static Connection getConnection() throws SQLException, IOException {
        if (connection == null || connection.isClosed()) {
            Properties props = new Properties();

            try (InputStream input = DatabaseConfig.class.getClassLoader().getResourceAsStream("config.properties")) {
                if (input == null) {
                    throw new IOException("Unable to find config.properties in the resources directory");
                }
                props.load(input);
            }
            String dbUrl = props.getProperty("db.url");
            if (dbUrl == null) {
                throw new IOException("db.url property not found in config.properties");
            }

            connection = DriverManager.getConnection(dbUrl);
        }
        return connection;
    }

    public void closeConnection() throws SQLException {
        if (connection != null && !connection.isClosed()) {
            connection.close();
        }
    }
}
